src/main/java/com/knighttodo/knighttodo/service/impl/BlockServiceImpl.java

```diff
@@ -70,34 +76,38 @@ public void deleteById(String blockId) {
     }
 
     private void fetchRoutines(BlockVO blockVO) {
-        List<RoutineVO> routines = routineService.findAllTemplates()
-            .stream().map(this::copyRoutine).collect(Collectors.toList());
-        blockVO.setRoutines(routines);
+        List<RoutineVO> routineVOS = routineService.findAllTemplates();
+        routineVOS.stream().map(this::copyRoutine).collect(Collectors.toList());
+        blockVO.setRoutines(routineVOS);
     }
 
     private RoutineVO copyRoutine(RoutineVO routineVO) {
-        RoutineVO routine = new RoutineVO();
-        copyTodosToRoutine(routine, routineVO);
-        routine.setTemplateId(routineVO.getTemplateId());
-        routine.setHardness(routineVO.getHardness());
-        routine.setScariness(routineVO.getScariness());
-        routine.setName(routineVO.getName());
-        routine.setBlock(routineVO.getBlock());
-        return routine;
+        RoutineVO updatedRoutineVO = new RoutineVO();
+        updatedRoutineVO.setHardness(routineVO.getHardness());
+        updatedRoutineVO.setScariness(routineVO.getScariness());
+        updatedRoutineVO.setName(routineVO.getName());
+        updatedRoutineVO.setBlock(routineVO.getBlock());
+        copyTodosToRoutine(updatedRoutineVO, routineVO);
+
+        updatedRoutineVO = routineService.save(routineVO.getBlock().getId(), updatedRoutineVO);
+        return updatedRoutineVO;
     }
 
-    private void copyTodosToRoutine(RoutineVO routine, RoutineVO routineVO) {
-        routine.setTodos(routineVO.getTodos().stream().map(this::copyTodo).collect(Collectors.toList()));
-        routine.getTodos().forEach(todoVO -> todoVO.setRoutine(routine));
+    private void copyTodosToRoutine(RoutineVO updatedRoutineVO, RoutineVO routineVO) {
+        updatedRoutineVO
+            .setTodos(routineVO.getTodos().stream().map(todo -> copyTodo(routineVO.getBlock().getId(), todo))
+                .collect(Collectors.toList()));
     }
 
-    private TodoVO copyTodo(TodoVO todoVO) {
-        TodoVO todo = new TodoVO();
-        todo.setTodoName(todoVO.getTodoName());
-        todo.setScariness(todoVO.getScariness());
-        todo.setHardness(todoVO.getHardness());
-        todo.setReady(false);
-        return todo;
+    private TodoVO copyTodo(String blockId, TodoVO todoVO) {
+        TodoVO updatedTodo = new TodoVO();
+        updatedTodo.setTodoName(todoVO.getTodoName());
+        updatedTodo.setScariness(todoVO.getScariness());
+        updatedTodo.setHardness(todoVO.getHardness());
+        updatedTodo.setReady(false);
+        updatedTodo.setRoutine(todoVO.getRoutine());
```

Comment:
Idk does it work but it seems like we are setting a reference to the incorrect routine (not copied, because and old todo should have ref to an old routine)

Reply:
It wasn't working right, it's nice that you understand it even without running the code)